barretenberg/cpp/src/barretenberg/vm2/common/aztec_types.hpp

```diff
@@ -294,11 +294,35 @@ struct TreeState {
     bool operator==(const TreeState& other) const = default;
 };
 
+struct InternalTreeSnapshots {
+    AppendOnlyTreeSnapshot l1ToL2MessageTree;
+    AppendOnlyTreeSnapshot noteHashTree;
+    AppendOnlyTreeSnapshot nullifierTree;
+    AppendOnlyTreeSnapshot publicDataTree;
+    AppendOnlyTreeSnapshot writtenPublicDataSlotsTree;
+
+    InternalTreeSnapshots() = default;
+
+    InternalTreeSnapshots(const TreeSnapshots& tree_snapshots,
+                          AppendOnlyTreeSnapshot written_public_data_slots_tree_snapshot)
+        : l1ToL2MessageTree(tree_snapshots.l1ToL2MessageTree)
+        , noteHashTree(tree_snapshots.noteHashTree)
+        , nullifierTree(tree_snapshots.nullifierTree)
+        , publicDataTree(tree_snapshots.publicDataTree)
+        , writtenPublicDataSlotsTree(written_public_data_slots_tree_snapshot)
+    {}
+
+    bool operator==(const InternalTreeSnapshots& other) const = default;
+};
+
 struct TreeStates {
     TreeState noteHashTree;
     TreeState nullifierTree;
     TreeState l1ToL2MessageTree;
-    TreeState publicDataTree;
+    // The public data tree counter is the size of the written public data slots tree minus one.
+    // Minus one due to the tree prefill.
+    AppendOnlyTreeSnapshot publicDataTree;
+    AppendOnlyTreeSnapshot writtenPublicDataSlotsTree;
```

Comment:
Let's see where this goes but in principle I'm a bit wary of mixing these trees. The first ones are actually protocol trees. The latter is an implementation detail due to having to count things in a circuit.

As I read this PR, my mental model is "this has to look like a dictionary/set to everything but the circuit", ideally. Let's see what we can do :)